### PR TITLE
[GLUTEN-4875][VL]Support spark sql conf sortBeforeRepartition to avoid stage partial retry casuing result mismatch

### PR DESCRIPTION
**Pull request by zjuwangg**
## What changes were proposed in this pull request?

Spark introduced `spark.sql.execution.sortBeforeRepartition` config in https://issues.apache.org/jira/browse/SPARK-23207 to keep the result correct, and we should do the same thing in gluten plan to achieve the same affact.

(Fixes: https://github.com/oap-project/gluten/issues/4875)

## How was this patch tested?

* Added ut to verify the executed plan is as expected.

### CODE REVIEW

File: cpp/core/jni/JniWrapper.cc [hunk 1]
```diff
@@ -830,7 +830,8 @@ JNIEXPORT jlong JNICALL Java_io_glutenproject_vectorized_ShuffleWriterJniWrapper
     jint startPartitionId,
     jint pushBufferMaxSize,
     jobject partitionPusher,
-    jstring partitionWriterTypeJstr) {
+    jstring partitionWriterTypeJstr,
+    jboolean sortBeforeRR) {
```
Comment (marin-ma) [inline cpp/core/jni/JniWrapper.cc:834]: Would the argument `sortBeforeRR` be used on the native side?

File: cpp/core/shuffle/RoundRobinPartitioner.h [hunk 2]
```diff
@@ -23,8 +23,8 @@ namespace gluten {
 
 class RoundRobinPartitioner final : public Partitioner {
  public:
-  RoundRobinPartitioner(int32_t numPartitions, int32_t startPartitionId)
-      : Partitioner(numPartitions, false), pidSelection_(startPartitionId % numPartitions) {}
+  RoundRobinPartitioner(int32_t numPartitions, int32_t startPartitionId, bool hasPid)
```
Comment (marin-ma) [inline cpp/core/shuffle/RoundRobinPartitioner.h:26]: `hasPid` should be `sortBeforeRR`, if we need this parameter.

File: cpp/core/jni/JniWrapper.cc (hunk 1 above)
Comment (zjuwangg) [reply cpp/core/jni/JniWrapper.cc:834]: Yes. RoundRobinPartitioner implement should know whether drop the first sorted column.

File: cpp/core/shuffle/RoundRobinPartitioner.h (hunk 2 above)
Comment (zjuwangg) [reply cpp/core/shuffle/RoundRobinPartitioner.h]: addressed.

File: cpp/core/jni/JniWrapper.cc (hunk 1 above)
Comment (marin-ma) [reply cpp/core/jni/JniWrapper.cc:834]: Would it be possible to add another projection before to drop the extra column?

Comment (zjuwangg) [reply cpp/core/jni/JniWrapper.cc:834]: It's possible, but we can leverage current design to eliminate  the first column and avoid the performance loss by providing extra project.

Comment (marin-ma) [reply cpp/core/jni/JniWrapper.cc:834]: I doubt that the impact of performance loss by adding one project which is only used to drop one column. The most part of extra overhead should come from the sort node. By adding this projection, we can also avoid adding argument `sortBeforeRR` to the several layers in the call stack.

Comment (zjuwangg) [reply cpp/core/jni/JniWrapper.cc:834]: Sounds great. I will address this comment later.

**Comment by zjuwangg** [comment]
@marin-ma plz help review again, there is a ci pre-check failure due to network problem.

File: gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/GlutenReplaceHashWithSortAggSuite.scala [hunk 3]
```diff
@@ -72,6 +72,7 @@ class GlutenReplaceHashWithSortAggSuite
                |   SORT BY key
                |)
                |GROUP BY key
+               |ORDER BY key
```
Comment (marin-ma) [inline gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/GlutenReplaceHashWithSortAggSuite.scala:75]: Why is this change needed? Could you add some comments?

File: gluten-ut/spark34/src/test/scala/org/apache/spark/sql/execution/GlutenReplaceHashWithSortAggSuite.scala [hunk 4]
```diff
@@ -71,6 +71,7 @@ class GlutenReplaceHashWithSortAggSuite
                |   SORT BY key
                |)
                |GROUP BY key
+               |ORDER BY key
```
Comment (marin-ma) [inline gluten-ut/spark34/src/test/scala/org/apache/spark/sql/execution/GlutenReplaceHashWithSortAggSuite.scala:74]: Same as above.

File: gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/GlutenReplaceHashWithSortAggSuite.scala (hunk 3 above)
Comment (zjuwangg) [reply gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/GlutenReplaceHashWithSortAggSuite.scala:75]: Because repartition modification causing the result sort order not same and casuing the result not same.

File: gluten-ut/spark34/src/test/scala/org/apache/spark/sql/execution/GlutenReplaceHashWithSortAggSuite.scala (hunk 4 above)
Comment (zjuwangg) [reply gluten-ut/spark34/src/test/scala/org/apache/spark/sql/execution/GlutenReplaceHashWithSortAggSuite.scala:74]: ditto

File: gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/GlutenReplaceHashWithSortAggSuite.scala (hunk 3 above)
Comment (marin-ma) [reply gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/GlutenReplaceHashWithSortAggSuite.scala:75]: Thanks. Please document in the code as this modification makes our test sql different from the original one https://github.com/apache/spark/blob/922f5f686dc72433a9028bbe471a35a5b84f2855/sql/core/src/test/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAggSuite.scala#L61-L72

**Comment by marin-ma** [comment]
LGTM. Please document the sql query modification in the code and I will proceed with merging. Thanks!

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/SparkPlanExecApiImpl.scala [hunk 5]
```diff
@@ -232,7 +232,24 @@ class SparkPlanExecApiImpl extends SparkPlanExecApi {
...
+        val dropSortColumnTransformer = ProjectExecTransformer(projectList.drop(1), sortByHashCode)
+        if (dropSortColumnTransformer.doValidate().isValid) {
+          ColumnarShuffleExchangeExec(
+            shuffle,
+            dropSortColumnTransformer,
+            dropSortColumnTransformer.output)
+        } else {
+          TransformHints.tagNotTransformable(
+            shuffle,
+            dropSortColumnTransformer.doValidate().reason.get)
```
Comment (ulysses-you) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/SparkPlanExecApiImpl.scala:250]: Can we reuse the previous validation result ?

Comment (zjuwangg) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/SparkPlanExecApiImpl.scala]: addressed

**Review by ulysses-you — APPROVED** [review]
